app/code/Magento/Catalog/Controller/Adminhtml/Product/Save.php

```diff
@@ -117,6 +117,7 @@ public function execute()
                 $productId = $product->getEntityId();
                 $productAttributeSetId = $product->getAttributeSetId();
                 $productTypeId = $product->getTypeId();
+                $this->getRequest()->setParam('type', $productTypeId);
```

Comment:
Modifying a request is not a good idea. This could be added to `$resultRedirect->setPath` calls but seems like it is not needed if you configure product attributes correctly.